design/backup-performance-improvements.md

```diff
@@ -0,0 +1,212 @@
+- The ItemBlock processing loop implemented in Phase 1 will be modified to send each newly-created ItemBlock to the shared channel rather than calling `BackupItemBlock` inline.
+- Users will be able to configure the number of workers available for concurrent `BackupItemBlock` operations.
+- Access to the BackedUpItems map must be synchronized
+
+## Detailed Design
+
+### Phase 1: ItemBlock processing
+
+#### BackupItemAction plugin changes
+
+In order for Velero to identify groups of items to back up together in an ItemBlock, we need a way to identify items which need to be backed up along with the current item. While the current `Execute` BackupItemAction method does return a list of additional items which are required by the current item, we need to know this *before* we start the item backup. To support this, we need a new API method, `GetAdditionalItems` which Velero will call on each item as it processes it for an ItemBlock. The expectation is that this method will return the same items as currently returned as additional items by the current `Execute` method, with the exception that items which are not created until calling `Execute` should not be returned here, as they don't exist yet.
+
+#### Proto changes (compiled into golang by protoc)
+
+The BackupItemAction service gets one new rpc method:
+```
+service BackupItemAction {
+    rpc GetAdditionalItems(BackupItemActionGetAdditionalItemsRequest) returns (BackupItemActionGetAdditionalItemsResponse);
+}
+```
+
+To support this new rpc method, we define new request/response message types:
+```
+message BackupItemActionAdditionalItemsRequest {
+    string plugin = 1;
+    bytes item = 2;
+    bytes backup = 3;
+}
+
+message BackupItemActionAdditionalItemsResponse {
+    repeated generated.ResourceIdentifier additionalItems = 1;
+}
+```
+
+A new PluginKind, `BackupItemActionV3`, will be created, and the backup process will be modified to use this plugin kind. Unlike with the V1->V2 transition, however, we will not provide a V3 adapter for BIAv2, because there is no reliable way to know what items to return from the new method. Instead, Velero will only invoke the new `GetAdditionalItems` method if *all* registered plugins are V3. If there are any V2 plugins registered, then we will continue to use the BIAv2 API, and process item blocks with only one item. If there are both V2 and V3 plugins, we will adapt (i.e. downgrade) V3 plugins to V2 rather than vice versa. In order for Velero to support item block processing out of the box, all existing V1 and V2 plugins in velero core and supported plugins will need to be converted to V3 as part of implementation. If we are adapting to V2 plugins, then velero will *not* call the new V3 plugin API method, and every item returned from the Item Collector will be in its own ItemBlock of one.
```

Comment:
There are a couple of probelems with 3) and 4)
The whole point of creating item blocks is to identify items that need to be backed up together. That's also the point of the v2 additional items return from Execute(), so any item returned by Execute() needs to be backed up with the current block, whether or not it was returned via a the new preprocessing method. This means a couple things:
1) if we nave non-v3 plugins, we have no way of knowing at block creation time whether already-existing items in the item collector list belong to a block, so by the time we get to step 3 above that item may already be in-process for backup in a parallel goroutine. For things like PVCs, this could lead to undesirable behavior
2) Regardless of plugin version, when Execute() returns an additional item, we need to back it up right now (using the current BackupItem logic) regardless of whether it's already in the item block. The BackedUpItems map prevents us from backing up  a single item twice.

Regarding 5), it's actually the reverse of that. Since the worker goroutines process ItemBlocks in parallel, it means that two items in different ItemBlocks may be processed at the same time. However, within an item block, items will always be processed in sequence, and because we are preserving the existing additional item processing within the BackupItem() func, within an item block, the sequence of backing up will be the same as today.

Reply:
My 1~5 (sorry for typo of 5, I have corrected it) above may not be sophisticated, however, if BIAv3 doesn't support to work with BIAV2, it will be a big break change problem, Velero especially the BIA is an open architecture, we don't know how many custom BIAs held by users/vendors, if we go in this way, I am afraid BIAv3 will not go full functioning in a very long time since it is hard to make all BIAs are BIAv3 in the user environments.

Reply:
Updated the design to reflect that it's no longer required that every BIA implement v3